src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioDocumentTrackerProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    internal class DefaultVisualStudioDocumentTrackerProvider : VisualStudioDocumentTrackerProvider
+    {
+        private readonly ForegroundDispatcher _dispatcher;
+
+        public DefaultVisualStudioDocumentTrackerProvider(ForegroundDispatcher dispatcher)
+        {
+            if (dispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(dispatcher));
+            }
+
+            _dispatcher = dispatcher;
+        }
+
+        public override VisualStudioDocumentTracker GetTracker(ITextView textView)
+        {
+            if (textView == null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            _dispatcher.AssertForegroundThread();
+
+            // While it's definitely possible to have multiple Razor text buffers attached to the same text view, there's
+            // no real scenario for it. This method always returns the tracker for the first Razor text buffer, but the
+            // other functionality for this class will maintain state correctly for the other buffers.
+            var textBuffer = textView.BufferGraph.GetRazorBuffers().FirstOrDefault();
+            if (textBuffer == null)
+            {
+                // No Razor buffer, nothing to track.
+                return null;
+            }
+
+            // A little bit of hardening here, to make sure our assumptions are correct.
+            if (!textBuffer.Properties.TryGetProperty(typeof(VisualStudioDocumentTracker), out VisualStudioDocumentTracker tracker))
```

Comment:
nit: needs more var

Reply:
lol could do `TryGetProperty<VisualStudioDocumentTracker>` to use `var` but doesn't buy anything imo. Properties bag is `object -> object`

Reply:
Ah,